src/main/java/com/flab/offcoupon/service/coupon_use/OrderService.java

```diff
@@ -0,0 +1,132 @@
+import com.flab.offcoupon.util.ResponseDTO;
+import lombok.RequiredArgsConstructor;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.time.LocalDateTime;
+import java.util.ArrayList;
+import java.util.List;
+
+import static com.flab.offcoupon.domain.entity.OrderCoupon.createOrderCoupon;
+import static com.flab.offcoupon.domain.entity.OrderDetail.createOrderDetail;
+import static com.flab.offcoupon.domain.entity.params.OrderInfo.createOrderInfo;
+import static com.flab.offcoupon.exception.coupon.CouponErrorMessage.COUPON_IS_NOT_ACTIVE;
+import static com.flab.offcoupon.exception.coupon.CouponErrorMessage.COUPON_USAGE_INVALID_PERIOD;
+
+@RequiredArgsConstructor
+@Service
+public class OrderService {
+
+    private final CouponIssueRepository couponIssueRepository;
+    private final CouponRepository couponRepository;
+    private final ProductRepository productRepository;
+    private final OrderDetailRepository orderDetailRepository;
+    private final OrderCouponRepository orderCouponRepository;
+
+    /**
+     * 사용 가능한 쿠폰 목록 조회
+     * @param memberId 회원 ID
+     * @param productId 상품 ID
+     * @param now 현재 날짜
+     * @return 사용 가능한 쿠폰 목록
+     */
+    @Transactional(readOnly = true)
+    public ResponseDTO<List<AvailableCouponsByMemberIdResponse>> getAvailableCoupons(final long memberId, final long productId, final LocalDateTime now) {
+        List<AvailableCouponsByMemberIdVo> availableCoupons =
+                couponIssueRepository.getAvailableCoupons(new MemberIdProductIdNowVo(memberId, productId, now));
+        List<AvailableCouponsByMemberIdResponse> responseList = new ArrayList<>();
+        filterAvailableCouponsWithMinPrice(productId, availableCoupons, responseList);
+        return ResponseDTO.getSuccessResult(responseList);
+    }
+
+    /**
+     * 최소 주문 금액과 비교하여 조건에 맞는 쿠폰만 결과 리스트에 추가합니다.
+     *
+     * @param productId        상품 ID
+     * @param availableCoupons 사용 가능한 쿠폰 목록
+     * @param responseList     결과 리스트
+     */
+    private void filterAvailableCouponsWithMinPrice(long productId, List<AvailableCouponsByMemberIdVo> availableCoupons, List<AvailableCouponsByMemberIdResponse> responseList) {
+        // min_price(상품의 최소 주문 금액)과 비교하여 조건에 맞는 쿠폰만 결과 리스트에 추가합니다.
+        if (!availableCoupons.isEmpty()) {
+            long accumulatedDiscountPrice = 0;
+            for (AvailableCouponsByMemberIdVo availableCoupon : availableCoupons) {
+                // 최소 주문 금액까지 할인 가능한 금액 누적
+                accumulatedDiscountPrice += availableCoupon.discountedPrice();
+                // 할인 금액이 min_price보다 작거나 같으면 결과 리스트에 추가
+                if (accumulatedDiscountPrice <= productRepository.getProductMinOrderPriceById(productId)) {
+                    AvailableCouponsByMemberIdResponse response = new AvailableCouponsByMemberIdResponse(availableCoupon);
+                    responseList.add(response);
+                }
+            }
+        }
+    }
+
```

Comment:
메서드 내부 코드의 depth가 상당히 깊습니다.
코드블록이 깊어지면 깊어질수록 이 코드가 어떤 조건일 때 수행되는지 판단하기 어렵습니다. 심해지면 (재밌는 짤이지만) 아래와 같은 코드가 나오기도 합니다. 아래와 같은 코드일 때 어떤 조건에서 이 코드가 수행되는지 알 수 있을까요?
![image](https://github.com/f-lab-edu/off-coupon/assets/131429812/6b1c32df-0a41-4ab7-97d8-45e23fd1da67)

그렇기 때문에 코드를 작성하실 때에는 코드가 중첩된상태가 되지 않도록 주의하며, 여러 기법을 사용해야 합니다.
그 기법중 하나는 early return pattern 입니다. 아래와 같이 하나의 코드 중첩을 없앨 수 있어요. 어떻게 하면 코드 깊이를 줄일 수 있을지 고민하며 한번 리팩토링해보세요~

```suggestion
    private void filterAvailableCouponsWithMinPrice(long productId, List<AvailableCouponsByMemberIdVo> availableCoupons, List<AvailableCouponsByMemberIdResponse> responseList) {
        // min_price(상품의 최소 주문 금액)과 비교하여 조건에 맞는 쿠폰만 결과 리스트에 추가합니다.
        if (availableCoupons.isEmpty()) {
          return List.of();
        }
           ...
        }
    

```

Reply:
인덴트를 줄이기 위해 람다표현식과 AtomicReference를 사용하는걸로 변경했습니다!